Kernel/Arch/aarch64/Interrupts.cpp

```diff
@@ -5,27 +5,132 @@
+static Array<GenericInterruptHandler*, 64> s_interrupt_handlers;
+
+extern "C" void handle_interrupt(TrapFrame const* const);
+extern "C" void handle_interrupt(TrapFrame const* const)
+{
+    for (auto& interrupt_controller : InterruptManagement::the().controllers()) {
+        auto pending_interrupts = interrupt_controller->pending_interrupts();
+
+        // TODO: Add these interrupts as a source of entropy for randomness.
+        u8 irq = 0;
+        while (pending_interrupts) {
+            if ((pending_interrupts & 0b1) != 0b1) {
+                irq += 1;
+                pending_interrupts >>= 1;
+                continue;
+            }
+
+            // TODO: Consider removing passing the RegisterState into the handle_interrupt()
+            //       function, since no IRQHandler seems to be using the registers.
+            RegisterState regs {};
+
+            auto* handler = s_interrupt_handlers[irq];
+            VERIFY(handler);
+            handler->increment_invoking_counter();
+            handler->handle_interrupt(regs);
+            handler->eoi();
+
+            irq += 1;
+            pending_interrupts >>= 1;
+        }
+    }
+}
+
+// FIXME: Share the code below with Arch/x86/common/Interrupts.cpp
+//        While refactoring, the interrupt handlers can also be moved into the InterruptManagement class.
+GenericInterruptHandler& get_interrupt_handler(u8 interrupt_number)
+{
+    auto*& handler_slot = s_interrupt_handlers[interrupt_number];
+    VERIFY(handler_slot != nullptr);
+    return *handler_slot;
+}
+
+static void revert_to_unused_handler(u8 interrupt_number)
 {
-    VERIFY_NOT_REACHED();
+    auto handler = new UnhandledInterruptHandler(interrupt_number);
+    handler->register_interrupt_handler();
 }
 
-void register_generic_interrupt_handler(u8, GenericInterruptHandler&)
+void register_generic_interrupt_handler(u8 interrupt_number, GenericInterruptHandler& handler)
 {
-    VERIFY_NOT_REACHED();
+    auto*& handler_slot = s_interrupt_handlers[interrupt_number];
+    if (handler_slot != nullptr) {
+        if (handler_slot->type() == HandlerType::UnhandledInterruptHandler) {
+            if (handler_slot) {
+                auto* unhandled_handler = static_cast<UnhandledInterruptHandler*>(handler_slot);
+                unhandled_handler->unregister_interrupt_handler();
+                delete unhandled_handler;
+            }
+            handler_slot = &handler;
+            return;
+        }
+        if (handler_slot->is_shared_handler() && !handler_slot->is_sharing_with_others()) {
+            VERIFY(handler_slot->type() == HandlerType::SharedIRQHandler);
+            static_cast<SharedIRQHandler*>(handler_slot)->register_handler(handler);
+            return;
+        }
+        if (!handler_slot->is_shared_handler()) {
+            if (handler_slot->type() == HandlerType::SpuriousInterruptHandler) {
+                // FIXME: Add support for spurious interrupts on aarch64
```

Comment:
Out of interest, what is a "spurious interrupt"? Like, a hw interrupt instead of a timer (?)

Reply:
Last time I checked a spurious interrupt was a interrupt without an actual cause....
But that seems not like what is meant here I guess

Reply:
This is one of the problems with sharing the x86 interrupt code, because apparently the PIC on x86 machines can cause a interrupt without there actually being a interrupt, thus being 'spurious': [PIC OSdev](https://wiki.osdev.org/8259_PIC#Spurious_IRQs)

On aarch64 this will never happen, which is why I also did not add the `SpuriousInterruptHandler` to the aarch64 build. But eventually this type should be removed from the general code, since it is a x86 specific thing.